pkg/ipam/ipallocator/allocator.go

```diff
@@ -138,30 +145,81 @@ func (a *SingleIPAllocator) AllocateIP(ip net.IP) error {
 	return nil
 }
 
+func (a *SingleIPAllocator) allocateOffset(i int) (net.IP, bool) {
+	if a.allocated.Bit(i) == 0 {
+		ip := utilnet.AddIPOffset(a.base, i)
+		if a.checkReserved(ip) != nil {
+			return nil, false
+		}
+		a.allocated.SetBit(a.allocated, i, 1)
+		a.count++
+		return ip, true
+	}
+
+	return nil, false
+}
+
 // AllocateNext allocates an IP from the IP range. It returns error if no IP is available.
 func (a *SingleIPAllocator) AllocateNext() (net.IP, error) {
 	a.mutex.Lock()
 	defer a.mutex.Unlock()
 	if a.count >= (a.max - len(a.reservedIPs)) {
 		return nil, fmt.Errorf("no available IP")
 	}
-	for i := 0; i < a.max; i++ {
-		if a.allocated.Bit(i) == 0 {
-			ip := utilnet.AddIPOffset(a.base, i)
-			if a.checkReserved(ip) != nil {
-				continue
-			}
-			a.allocated.SetBit(a.allocated, i, 1)
-			a.count++
+	for i := 0; i <= a.max; i++ {
+		if ip, ok := a.allocateOffset(i); ok {
 			return ip, nil
 		}
 	}
+
+	// we should never reach here
 	return nil, fmt.Errorf("no available IP")
 }
 
+// IP space starting from startIP, if specified. If continuus chunk is not
+// available, other available IPs will be allocated.
+// With bestEffort=true, AllocateChunk will succeed even if not enough IPs were
+// available for allocation. bestEffort=false functionality is used in MultiIPAllocator
+// to ensure race-safe allocation without the need to introduce an extra mutex.
+func (a *SingleIPAllocator) AllocateChunk(startIP net.IP, chunkSize int, bestEffort bool) ([]net.IP, error) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	if (a.count+chunkSize > (a.max - len(a.reservedIPs))) && !bestEffort {
+		return nil, fmt.Errorf("not enough available IPs")
+	}
+
+	var ips []net.IP
+
+	allocateInRange := func(min, max int) {
+		for i := min; (i < max) && (chunkSize-len(ips)) > 0; i++ {
+			if ip, ok := a.allocateOffset(i); ok {
```

Comment:
Do we need to check the difference between `allocate` and `preallocate`?

Reply:
`ipallocator` is not aware of this distinction, do you think it should be?

Reply:
I think it is OK if a "preallocated" IP is recognized as "allocated" by ipallocator as current code implementation, since agent can find it by StatefulSet name and index directly, and there is no need to allocate a preallocated IP by ipallocator.
https://github.com/antrea-io/antrea/blob/main/pkg/ipam/poolallocator/allocator.go#L101
@jianjuns @tnqn Do you think this is acceptable?